lexers/nix.lua

```diff
@@ -0,0 +1,81 @@
+-- Nix LPEG lexer.
+-- Written by Samuel Marquis.
+
+local lexer = lexer
+local P, S = lpeg.P, lpeg.S
+
+local lex = lexer.new(...)
+
+-- Keywords.
+lex:add_rule('keyword', lex:tag(lexer.KEYWORD, lex:word_match(lexer.KEYWORD)))
+
+-- Functions.
+lex:add_rule('function', lex:tag(lexer.FUNCTION_BUILTIN, lex:word_match(lexer.FUNCTION_BUILTIN)))
+
+-- Constants.
+lex:add_rule('constant', lex:tag(lexer.CONSTANT_BUILTIN, lex:word_match(lexer.CONSTANT_BUILTIN)))
+
+-- Strings.
+local str = lexer.range('"', true)
+local ml_str = lexer.range("''", false)
+lex:add_rule('string', lex:tag(lexer.STRING, str + ml_str))
+
+-- Paths.
+local path_char = lexer.alnum + S('_-.+')
+local path_seg = ('/' * path_char^1)
+local path = P('~')^-1 * path_char^0 * path_seg^1 * lpeg.P('/')^-1
+lex:add_rule('path', lex:tag(lexer.LINK, path))
+
+-- URIs.
+local uri_char = lexer.alnum + S("%/?:@&=+$,-_.!~*'")
+local uri = lexer.alpha * (lexer.alnum + S('+-.'))^0 * ':' * uri_char^1
+lex:add_rule('uri', lex:tag(lexer.LINK, uri))
+
+-- Angle-bracket paths.
+local spath = '<' * path_char^1 * path_seg^0 * '>'
+lex:add_rule('spath', lex:tag(lexer.LINK, spath))
+
+-- Identifiers.
+local id = (lexer.alpha + '_') * (lexer.alnum + S('_-'))^0
+lex:add_rule('identifier', lex:tag(lexer.IDENTIFIER, id))
+
+-- Comments.
+local line_comment = lexer.to_eol('#', true)
+local block_comment = lexer.range('/*', '*/')
+lex:add_rule('comment', lex:tag(lexer.COMMENT, line_comment + block_comment))
+
+-- Numbers.
+lex:add_rule('number', lex:tag(lexer.NUMBER, lexer.number))
+
+-- Operators.
+local ops = S('?+-.*/!<>=,;:()[]{}') + lex:word_match(lexer.OPERATOR)
+lex:add_rule('operator', lex:tag(lexer.OPERATOR, ops))
+
+-- Fold points.
+lex:add_fold_point(lexer.OPERATOR, '(', ')')
+lex:add_fold_point(lexer.OPERATOR, '[', ']')
+lex:add_fold_point(lexer.OPERATOR, '{', '}')
+lex:add_fold_point(lexer.COMMENT, '/*', '*/')
+
+-- Word lists.
+lex:set_word_list(lexer.KEYWORD, {
+	'if', 'then', 'else', 'assert', 'with',
+	'let', 'in', 'rec', 'inherit', 'or', '...',
+})
+
+lex:set_word_list(lexer.CONSTANT_BUILTIN, {
+	'builtins', 'true', 'false', 'null',
+})
+
+-- Directly accessible functions.
+lex:set_word_list(lexer.FUNCTION_BUILTIN, {
+	'derivation', 'import', 'abort', 'throw',
```

Comment:
When you say there are a ton of Nix builtin functions and contants, how many are we talking? I can see value in having them if they are part of the language.

For example, the C++ lexer includes over 100 stl functions and types: https://github.com/orbitalquark/scintillua/blob/default/lexers/cpp.lua. It may not be exhaustive, but it's quite helpful for editing a wide range of C++ code.

Reply:
The list of constants is [here](https://nix.dev/manual/nix/2.18/language/builtin-constants) and functions [here](https://nix.dev/manual/nix/2.18/language/builtins). I'll see if I can find a way to automate dropping all of them into a list, because the prospect of hand-copying them seems quite tedious--but I agree it'd be valuable to have them.

Reply:
Yeah, I'd like to have them if it's not too much trouble. Hopefully there's an automated way to do it. Thanks!

Reply:
Got all the builtins, probably going to come back to this in the future and add everything in `pkgs` and `pkgs.libs`/`pkgs.stdenv`, but those docs pages were a lot less trivial to scrape

Reply:
Okay, however now the lexer will highlight bare `add()` as a function, when it should be `builtins.add()`, right?

You'll have to modify your `function` rule to use two word lists, the first one is for non-builtins, and the second one is for builtins. For example, `lex:add_rule('function', lex:tag(lexer.FUNCTION_BUILTIN, lex:word_match(lexer.FUNCTION_BUILTIN) + 'builtins.' * lex:word_match(lexer.FUNCTION_BUILTIN .. '.builtins'))` where `lex:set_word_list(lexer.FUNCTION_BUILTIN .. '.builtins', { ... }` contains `'add'`, etc.

If you're tired of working on this I can take over when I have some time. I think I have enough context to push this over the finish line.

Reply:
No problem, I'll add this in a moment

Reply:
It does seem mildly sub-ideal to consume the `.` in `builtins.` so I'll see about matching without consuming; that way `builtins` can still match as a constant in its own right when it's used to get a function.

Reply:
On the contrary, I would like it to consume `builtins.` as part of the function, just like the Lua lexer does with things like `string.find()`.

Reply:
So would, for example, `builtins.currentSystem` consume the entire thing as a constant? This feels slightly dubious to me wrt `builtins` getting parsed two different ways and losing the `.` operator, but I'll defer if this desired.

Reply:
Yes, `builtins.currentSystem` would be all highlighted as a constant, including the `.` operator. The primary reason is for backtracking. In case clients want to start syntax highlighting from the current position (e.g. for performance), backtracking to a style change is sufficient. If we broke up `builtins.currentSystem`, then a client would backtrack to just the `.`, would start highlighting, but would not see `currentSystem` as belonging to `builtins`, and would highlight incorrectly. By highlighting it all, any backtracking would allow highlighting to start at `builtins`, catching any changes at the current position. I hope this makes sense.

Does Nix allow using `builtins` by itself? (e.g. passing as a parameter) If it doesn't, then I probably wouldn't highlight `builtins` by itself since it may be meaningless and give that hint to the user.

Reply:
I have yet to see it in the wild... but every day I encounter a new strange usage of this language that is foreign to me. I'll do a little digging but we can omit it for now, I'll make this and the other change

Reply:
Didn't find anything doing this in my cursory search. If I spot it down the line, I'll make another short PR for it.

Reply:
One more thought about this--people pretty frequently will import specific names they plan on using from `builtins` to their environment, and imo those should still get lexed, so I'll throw a `^0` on the `P('builtins.')`

Reply:
Just now realizing this might make the backtracking thing moot, as--to my understanding of what you've written at least--a highlight of `currentSystem` _would_ still be correct given how often that name & others like it are imported.